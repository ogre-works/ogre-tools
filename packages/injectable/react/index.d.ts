/// <reference types="react" />
import type {
  DiContainer,
  DiContainerForInjection,
} from '@ogre-tools/injectable';
import { IComputedValue } from 'mobx';

interface DiContainerProviderProps {
  di: DiContainer | DiContainerForInjection;
}

export const DiContextProvider: React.Provider<DiContainerProviderProps>;

export interface WithInjectablesSyncOptions<
  Dependencies extends object,
  Props extends object,
> {
  getProps: (di: DiContainer, props: Props) => Props & Dependencies;
}

export interface WithInjectablesAsyncOptions<
  Dependencies extends object,
  Props extends object,
> {
  getProps: (di: DiContainer, props: Props) => Promise<Props & Dependencies>;
  getPlaceholder: React.FunctionComponent<Props>;
}

export interface WithInjectables {
  <Dependencies extends object, Props extends object = {}>(
    Component: React.ElementType<Dependencies & Props>,
    options: WithInjectablesSyncOptions<Dependencies, Props>,
  ): React.FunctionComponent<Props>;

  <Dependencies extends object, Props extends object, Ref extends object>(
    Component: React.ForwardRefExoticComponent<
      Dependencies & Props & React.RefAttributes<Ref>
    >,
    options: WithInjectablesSyncOptions<Dependencies, Props>,
  ): React.ForwardRefExoticComponent<Props & React.RefAttributes<Ref>>;

  <Dependencies extends object, Props extends object = {}>(
    Component: React.ElementType<Dependencies & Props>,
    options: WithInjectablesAsyncOptions<Dependencies, Props>,
  ): React.FunctionComponent<Props>;

  <Dependencies extends object, Props extends object, Ref extends object>(
    Component: React.ForwardRefExoticComponent<
      Dependencies & Props & React.RefAttributes<Ref>
    >,
    options: WithInjectablesAsyncOptions<Dependencies, Props>,
  ): React.ForwardRefExoticComponent<Props & React.RefAttributes<Ref>>;
}

export const withInjectables: WithInjectables;

export type IAsyncComputed<T> = {
  value: IComputedValue<T>;
  pending: IComputedValue<boolean>;
  invalidate: () => void;
};

type AsyncComputedParams<T> = {
  getValueFromObservedPromise: () => Promise<T>;
  valueWhenPending?: T;
  betweenUpdates?: 'show-pending-value' | 'show-latest-value';
};

export function asyncComputed<T>(
  configuration: AsyncComputedParams<T>,
): IAsyncComputed<T>;

export function registerInjectableReact(di: DiContainer): void;
