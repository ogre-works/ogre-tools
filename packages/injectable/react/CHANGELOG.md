# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

### [15.1.2](https://github.com/ogre-works/ogre-tools/compare/v15.1.1...v15.1.2) (2023-03-02)

**Note:** Version bump only for package @ogre-tools/injectable-react

### [15.1.1](https://github.com/ogre-works/ogre-tools/compare/v15.1.0...v15.1.1) (2023-02-27)

**Note:** Version bump only for package @ogre-tools/injectable-react

## [15.1.0](https://github.com/ogre-works/ogre-tools/compare/v15.0.1...v15.1.0) (2023-02-22)

### Features

- Make placeholder of withInjectables able to be more specialized by passing props to it ([d3fa8aa](https://github.com/ogre-works/ogre-tools/commit/d3fa8aa8528533942c612cf92cf54cad70ad212d))

### [15.0.1](https://github.com/ogre-works/ogre-tools/compare/v15.0.0...v15.0.1) (2023-02-21)

### Bug Fixes

- Fix bad import path ([c90d54b](https://github.com/ogre-works/ogre-tools/commit/c90d54b4c2713bccff6e2830808994a4c7a05f77))

## [15.0.0](https://github.com/ogre-works/ogre-tools/compare/v14.0.3...v15.0.0) (2023-02-20)

### ⚠ BREAKING CHANGES

- `injectable-react` now requires `registerInjectableReact(di)` to work.

### Features

- Remove too complex concept of "ad-hoc"-injectables (ie. injectable without registration) ([3baaf04](https://github.com/ogre-works/ogre-tools/commit/3baaf049bff802f015c5dc8497aa6c3e01f91cfe))

### [14.0.3](https://github.com/ogre-works/ogre-tools/compare/v14.0.2...v14.0.3) (2022-11-23)

**Note:** Version bump only for package @ogre-tools/injectable-react

### [14.0.2](https://github.com/ogre-works/ogre-tools/compare/v14.0.1...v14.0.2) (2022-11-23)

**Note:** Version bump only for package @ogre-tools/injectable-react

### [14.0.1](https://github.com/ogre-works/ogre-tools/compare/v14.0.0...v14.0.1) (2022-11-23)

**Note:** Version bump only for package @ogre-tools/injectable-react

## [14.0.0](https://github.com/ogre-works/ogre-tools/compare/v13.2.1...v14.0.0) (2022-11-22)

**Note:** Version bump only for package @ogre-tools/injectable-react

### [13.2.1](https://github.com/ogre-works/ogre-tools/compare/v13.2.0...v13.2.1) (2022-11-17)

**Note:** Version bump only for package @ogre-tools/injectable-react

## [13.2.0](https://github.com/ogre-works/ogre-tools/compare/v13.1.0...v13.2.0) (2022-11-17)

**Note:** Version bump only for package @ogre-tools/injectable-react

## [13.1.0](https://github.com/ogre-works/ogre-tools/compare/v13.0.0...v13.1.0) (2022-11-15)

**Note:** Version bump only for package @ogre-tools/injectable-react

## [13.0.0](https://github.com/ogre-works/ogre-tools/compare/v12.0.1...v13.0.0) (2022-11-14)

**Note:** Version bump only for package @ogre-tools/injectable-react

### [12.0.1](https://github.com/ogre-works/ogre-tools/compare/v12.0.0...v12.0.1) (2022-10-25)

### Bug Fixes

- Manually update versions of peerDependencies as Lerna doesn't do it for some reason ([0b46667](https://github.com/ogre-works/ogre-tools/commit/0b46667c106a93131f6320222aced3187cd9a292))

## [12.0.0](https://github.com/ogre-works/ogre-tools/compare/v11.0.0...v12.0.0) (2022-10-25)

### ⚠ BREAKING CHANGES

- Make asyncComputed receive parameters as value object instead of many arguments

### Features

- Make asyncComputed able to show latest value (instead of pending value) between updates ([d543053](https://github.com/ogre-works/ogre-tools/commit/d5430534f16b67d7b5f09a5ce07d893dab48426d))
- Make asyncComputed receive parameters as value object instead of many arguments ([7cf0035](https://github.com/ogre-works/ogre-tools/commit/7cf00350e079a65d8aa9d904d71fb07ce0eff520))

## [11.0.0](https://github.com/ogre-works/ogre-tools/compare/v10.3.2...v11.0.0) (2022-10-04)

**Note:** Version bump only for package @ogre-tools/injectable-react

### [10.3.2](https://github.com/ogre-works/ogre-tools/compare/v10.3.1...v10.3.2) (2022-09-26)

**Note:** Version bump only for package @ogre-tools/injectable-react

### [10.3.1](https://github.com/ogre-works/ogre-tools/compare/v10.3.0...v10.3.1) (2022-09-26)

**Note:** Version bump only for package @ogre-tools/injectable-react

## [10.3.0](https://github.com/ogre-works/ogre-tools/compare/v10.2.0...v10.3.0) (2022-09-26)

### Features

- Make build of a ts-package include type information ([d174431](https://github.com/ogre-works/ogre-tools/commit/d1744317a55a4975f68452534321d98a8ff5e543))

## [10.2.0](https://github.com/ogre-works/ogre-tools/compare/v10.1.0...v10.2.0) (2022-09-23)

**Note:** Version bump only for package @ogre-tools/injectable-react

## [10.1.0](https://github.com/ogre-works/ogre-tools/compare/v10.0.0...v10.1.0) (2022-09-02)

**Note:** Version bump only for package @ogre-tools/injectable-react

## [10.0.0](https://github.com/ogre-works/ogre-tools/compare/v9.0.3...v10.0.0) (2022-09-01)

**Note:** Version bump only for package @ogre-tools/injectable-react

### [9.0.3](https://github.com/ogre-works/ogre-tools/compare/v9.0.2...v9.0.3) (2022-08-25)

### Bug Fixes

- Make injectable-react not break React reconciliation when using async dependencies ([049c84a](https://github.com/ogre-works/ogre-tools/commit/049c84a022ed18e14cd6b1d0a8ad7358ccec8640))

### [9.0.2](https://github.com/ogre-works/ogre-tools/compare/v9.0.1...v9.0.2) (2022-08-10)

**Note:** Version bump only for package @ogre-tools/injectable-react

### [9.0.1](https://github.com/ogre-works/ogre-tools/compare/v9.0.0...v9.0.1) (2022-06-29)

**Note:** Version bump only for package @ogre-tools/injectable-react

## [9.0.0](https://github.com/ogre-works/ogre-tools/compare/v8.0.0...v9.0.0) (2022-06-20)

### ⚠ BREAKING CHANGES

- Make containers have ID for better error messages

### Features

- Make containers have ID for better error messages ([ea3ac4c](https://github.com/ogre-works/ogre-tools/commit/ea3ac4c4becbb79509ffe19f66ffac62364e22d6))

## [8.0.0](https://github.com/ogre-works/ogre-tools/compare/v7.1.0...v8.0.0) (2022-06-15)

### Reverts

- Revert "chore: Consolidate to relying of peer dependencies in injectable extensions" ([8f9b9b2](https://github.com/ogre-works/ogre-tools/commit/8f9b9b2afee9154e523254ea0d9338aa5fdd4fb4))

## [7.1.0](https://github.com/ogre-works/ogre-tools/compare/v7.0.0...v7.1.0) (2022-06-09)

**Note:** Version bump only for package @ogre-tools/injectable-react

## [7.0.0](https://github.com/ogre-works/ogre-tools/compare/v6.0.1...v7.0.0) (2022-04-27)

### ⚠ BREAKING CHANGES

- Extract auto-registration as extension

### Miscellaneous Chores

- Extract auto-registration as extension ([8395a7a](https://github.com/ogre-works/ogre-tools/commit/8395a7a8eba457f98ebd10a1dfa7c6cf901fdc34))

### [6.0.1](https://github.com/ogre-works/ogre-tools/compare/v6.0.0...v6.0.1) (2022-03-30)

**Note:** Version bump only for package @ogre-tools/injectable-react

## [6.0.0](https://github.com/ogre-works/ogre-tools/compare/v5.2.0...v6.0.0) (2022-03-29)

**Note:** Version bump only for package @ogre-tools/injectable-react

## [5.2.0](https://github.com/ogre-works/ogre-tools/compare/v5.1.3...v5.2.0) (2022-03-23)

**Note:** Version bump only for package @ogre-tools/injectable-react

### [5.1.3](https://github.com/ogre-works/ogre-tools/compare/v5.1.2...v5.1.3) (2022-03-18)

**Note:** Version bump only for package @ogre-tools/injectable-react

### [5.1.2](https://github.com/ogre-works/ogre-tools/compare/v5.1.1...v5.1.2) (2022-03-14)

**Note:** Version bump only for package @ogre-tools/injectable-react

### [5.1.1](https://github.com/ogre-works/ogre-tools/compare/v5.1.0...v5.1.1) (2022-03-11)

**Note:** Version bump only for package @ogre-tools/injectable-react

## [5.1.0](https://github.com/ogre-works/ogre-tools/compare/v5.0.0...v5.1.0) (2022-03-11)

### Features

- Add injection context for withInjectables ([d69a57d](https://github.com/ogre-works/ogre-tools/commit/d69a57d7d1428342b76bb0d6ef272ede6c85858c))

### Bug Fixes

- Resolve PR comments ([bb2e1de](https://github.com/ogre-works/ogre-tools/commit/bb2e1debdcfe901f998eabcaba1941222e917950))
- **typings:** Improve typings to work with arbitrary injection params ([c1d900a](https://github.com/ogre-works/ogre-tools/commit/c1d900a22ae9d609e3b70b3d0a034dd5e81901b0))

## [5.0.0](https://github.com/ogre-works/ogre-tools/compare/v4.1.0...v5.0.0) (2022-02-09)

### ⚠ BREAKING CHANGES

- Consolidate typing of injectable and injectable-react

### Bug Fixes

- Consolidate typing of injectable and injectable-react ([a52180d](https://github.com/ogre-works/ogre-tools/commit/a52180d28119e544c5023a8706ca2a077f2217cf))

## [4.1.0](https://github.com/ogre-works/ogre-tools/compare/v4.0.0...v4.1.0) (2022-02-08)

**Note:** Version bump only for package @ogre-tools/injectable-react

## [4.0.0](https://github.com/ogre-works/ogre-tools/compare/v3.2.1...v4.0.0) (2022-02-07)

### Bug Fixes

- fix build by using webpack made available to a package by lerna ([5b08e24](https://github.com/ogre-works/ogre-tools/commit/5b08e2472fe06514901546e8a5eb8d8664282a0c))
