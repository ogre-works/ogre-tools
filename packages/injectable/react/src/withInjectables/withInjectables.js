import React from 'react';
import { constant } from 'lodash/fp';
import { Observer } from 'mobx-react';
import { action, observable } from 'mobx';

import { isPromise } from '@lensapp/fp';
import { getInjectable, lifecycleEnum } from '@lensapp/injectable';

const { Provider: DiContextProvider, Consumer: DiContextConsumer } =
  React.createContext();

export { DiContextProvider };

export const componentNameMapInjectable = getInjectable({
  id: 'component-name-map',
  instantiate: () => new Map(),
});

export default (Component, { getPlaceholder = constant(null), getProps }) =>
  React.memo(
    React.forwardRef((props, ref) => (
      <DiContextConsumer>
        {({ di }) => {
          const componentNameMap = di.inject(componentNameMapInjectable);

          if (!componentNameMap.has(Component)) {
            componentNameMap.set(
              Component,
              Component.displayName ||
                Component.name ||
                `anonymous-component-${componentNameMap.size}`,
            );
          }

          const componentContext = {
            injectable: {
              id: componentNameMap.get(Component),
              lifecycle: lifecycleEnum.transient,
            },
          };

          const diForComponentContext = {
            ...di,

            inject: (alias, parameter) =>
              di.inject(alias, parameter, componentContext),

            injectMany: (injectionToken, parameter) =>
              di.injectMany(injectionToken, parameter, componentContext),
          };

          const maybeAsyncProps = getProps(diForComponentContext, props);
          const refProps = ref ? { ref } : {};

          if (!isPromise(maybeAsyncProps)) {
            return <Component {...refProps} {...maybeAsyncProps} />;
          }

          const observablePropsPromise = getObservablePromise(maybeAsyncProps);

          return (
            <Observer>
              {() => {
                const syncProps = observablePropsPromise.value;

                if (!syncProps) {
                  return getPlaceholder(props);
                }

                return <Component {...refProps} {...syncProps} />;
              }}
            </Observer>
          );
        }}
      </DiContextConsumer>
    )),
  );

const getObservablePromise = asyncValue => {
  const observableObject = observable({ value: null }, undefined, {
    deep: false,
  });

  asyncValue.then(
    action(syncValue => {
      observableObject.value = syncValue;
    }),
  );

  return observableObject;
};
