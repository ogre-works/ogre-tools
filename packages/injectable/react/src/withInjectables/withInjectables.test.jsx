import React from 'react';
import { setImmediate as flushMicroTasks } from 'timers';
import { render } from '@testing-library/react';
import { act } from '@testing-library/react';

import {
  createContainer,
  getInjectable,
  getInjectionToken,
  lifecycleEnum,
} from '@ogre-tools/injectable';

import withInjectables, { DiContextProvider } from './withInjectables';
import asyncFn from '@async-fn/jest';
import { observable, runInAction } from 'mobx';
import { observer } from 'mobx-react';
import registerInjectableReact from '../registerInjectableReact/registerInjectableReact';

const flushPromises = () => new Promise(flushMicroTasks);

const mountFor =
  di =>
  (node, ...rest) =>
    render(<DiContextProvider value={{ di }}>{node}</DiContextProvider>, {
      ...rest,
    });

describe('withInjectables', () => {
  let di;
  let mount;

  beforeEach(() => {
    di = createContainer('some-container');

    registerInjectableReact(di);

    mount = mountFor(di);
  });

  it('given async dependency, placeholder and parent component, when parent component rerenders, does not remount component with the dependency', async () => {
    const componentDidMountMock = jest.fn();

    const injectable = getInjectable({
      id: 'some-injectable-id',
      instantiate: async () => await Promise.resolve('some-injectable-value'),
    });

    di.register(injectable);

    class DumbTestComponent extends React.Component {
      componentDidMount() {
        componentDidMountMock();
      }

      render() {
        let { someDependency, ...props } = this.props;

        return <div {...props}>Some content: "{someDependency}"</div>;
      }
    }

    const SmartTestComponent = withInjectables(DumbTestComponent, {
      getPlaceholder: () => <div />,

      getProps: async (di, props) => ({
        someDependency: await di.inject(injectable),
        ...props,
      }),
    });

    const SomeParentComponent = observer(({ someObservable }) => (
      <div>
        {someObservable.get()}

        <SmartTestComponent data-some-prop-test />
      </div>
    ));

    const someObservable = observable.box('some-value');

    mount(<SomeParentComponent someObservable={someObservable} />);

    await act(() => {
      runInAction(() => {
        someObservable.set('some-other-value');
      });
    });

    expect(componentDidMountMock).toHaveBeenCalledTimes(1);
  });

  it('given sync dependency and parent component, when parent component rerenders, does not remount component with the dependency', () => {
    const componentDidMountMock = jest.fn();

    const injectable = getInjectable({
      id: 'some-injectable-id',
      instantiate: () => 'some-injectable-value',
    });

    di.register(injectable);

    class DumbTestComponent extends React.Component {
      componentDidMount() {
        componentDidMountMock();
      }

      render() {
        let { someDependency, ...props } = this.props;

        return <div {...props}>Some content: "{someDependency}"</div>;
      }
    }

    const SmartTestComponent = withInjectables(DumbTestComponent, {
      getProps: (di, props) => ({
        someDependency: di.inject(injectable),
        ...props,
      }),
    });

    const SomeParentComponent = observer(({ someObservable }) => (
      <div>
        {someObservable.get()}

        <SmartTestComponent data-some-prop-test />
      </div>
    ));

    const someObservable = observable.box('some-value');

    mount(<SomeParentComponent someObservable={someObservable} />);

    act(() => {
      runInAction(() => {
        someObservable.set('some-other-value');
      });
    });

    expect(componentDidMountMock).toHaveBeenCalledTimes(1);
  });

  it('given component and sync dependencies, when rendered, renders with dependencies', () => {
    const injectable = getInjectable({
      id: 'some-injectable-id',

      lifecycle: lifecycleEnum.transient,

      instantiate: () => 'some-injectable-value',
    });

    di.register(injectable);

    const DumbTestComponent = ({ someDependency, ...props }) => (
      <div {...props}>Some content: "{someDependency}"</div>
    );

    const SmartTestComponent = withInjectables(DumbTestComponent, {
      getProps: (di, props) => ({
        someDependency: di.inject(injectable),
        ...props,
      }),
    });

    const rendered = mount(<SmartTestComponent data-some-prop-test />);

    expect(rendered.baseElement).toMatchSnapshot();
  });

  it('given component and sync dependencies, when rendered without provider, renders to null', () => {
    const injectable = getInjectable({
      id: 'some-injectable-id',

      lifecycle: lifecycleEnum.transient,

      instantiate: () => 'some-injectable-value',
    });

    di.register(injectable);

    const DumbTestComponent = ({ someDependency, ...props }) => (
      <div {...props}>Some content: "{someDependency}"</div>
    );

    const SmartTestComponent = withInjectables(DumbTestComponent, {
      getProps: (di, props) => ({
        someDependency: di.inject(injectable),
        ...props,
      }),
    });

    const rendered = render(<SmartTestComponent data-some-prop-test />);

    expect(rendered.baseElement).toMatchSnapshot();
  });

  [
    {
      name: 'given anonymous component and a dependency cycle, when rendered, throws context for a uniquely generated component name and cycle',
      getComponent: () => () => 'irrelevant',
      injectOrInjectMany: 'inject',

      expectedError:
        'Cycle of injectables encountered: "some-container" -> "anonymous-component-0" -> "some-injectable-id" -> "some-other-injectable-id" -> "some-injectable-id"',
    },

    {
      name: 'given named component and a dependency cycle, when rendered, throws context for the component and cycle',

      getComponent: () =>
        function SomeNamedComponent() {
          return 'irrelevant';
        },

      injectOrInjectMany: 'inject',

      expectedError:
        'Cycle of injectables encountered: "some-container" -> "SomeNamedComponent" -> "some-injectable-id" -> "some-other-injectable-id" -> "some-injectable-id"',
    },

    {
      name: 'given component with display name and a dependency cycle, when rendered, throws context for the component and cycle',

      getComponent: () => {
        const Component = () => 'irrelevant';
        Component.displayName = 'some-component-display-name';
        return Component;
      },

      injectOrInjectMany: 'inject',

      expectedError:
        'Cycle of injectables encountered: "some-container" -> "some-component-display-name" -> "some-injectable-id" -> "some-other-injectable-id" -> "some-injectable-id"',
    },

    {
      name: 'given class component and a dependency cycle, when rendered, throws context for the component and cycle',

      getComponent: () =>
        class SomeClassComponent extends React.Component {
          render() {
            return 'irrelevant';
          }
        },

      injectOrInjectMany: 'inject',

      expectedError:
        'Cycle of injectables encountered: "some-container" -> "SomeClassComponent" -> "some-injectable-id" -> "some-other-injectable-id" -> "some-injectable-id"',
    },

    {
      name: 'given component, a dependency cycle and injecting a token, when rendered, throws context for the component and cycle',

      getComponent: () => () => 'irrelevant',

      injectUsing: 'injectionToken',

      expectedError:
        'Cycle of injectables encountered: "some-container" -> "anonymous-component-0" -> "some-injection-token" -> "some-injectable-id" -> "some-other-injectable-id" -> "some-injectable-id"',
    },
  ].forEach(scenario => {
    xit(scenario.name, () => {
      const someInjectionToken = getInjectionToken({
        id: 'some-injection-token',
      });

      const injectable = getInjectable({
        id: 'some-injectable-id',
        instantiate: di => di.inject(otherInjectable),
        injectionToken: someInjectionToken,
      });

      const otherInjectable = getInjectable({
        id: 'some-other-injectable-id',
        instantiate: di => di.inject(injectable),
      });

      di.register(injectable, otherInjectable);

      const DumbTestComponent = scenario.getComponent();

      const SmartTestComponent = withInjectables(DumbTestComponent, {
        getProps: (di, props) => ({
          someDependency:
            scenario.injectUsing === 'injectionToken'
              ? di.injectMany(someInjectionToken)
              : di.inject(injectable),
          ...props,
        }),
      });

      const onErrorMock = jest.fn();

      const storedConsoleError = console.error;
      console.error = () => {};

      mount(
        <ErrorBoundary onError={onErrorMock}>
          <SmartTestComponent data-some-prop-test />
        </ErrorBoundary>,
      );

      console.error = storedConsoleError;

      expect(onErrorMock).toHaveBeenCalledWith(scenario.expectedError);
    });
  });

  it('given anonymous component, injecting a token, when rendered, works', () => {
    const someInjectionToken = getInjectionToken({
      id: 'some-injection-token',
    });

    const injectable = getInjectable({
      id: 'some-injectable-id',
      instantiate: () => 42,
      injectionToken: someInjectionToken,
    });

    di.register(injectable);

    const SmartTestComponent = withInjectables(() => 'irrelevant', {
      getProps: (di, props) => ({
        someDependency: di.injectMany(someInjectionToken),
        ...props,
      }),
    });

    const rendered = mount(<SmartTestComponent data-some-prop-test />);

    expect(rendered.baseElement).toMatchSnapshot();
  });

  it('given component, and rendered, when re-rendered, works', () => {
    const someObservable = observable.box(24);

    const injectable = getInjectable({
      id: 'some-injectable-id',
      instantiate: () => someObservable,
    });

    di.register(injectable);

    const SmartTestComponent = withInjectables(() => 'irrelevant', {
      getProps: (di, props) => ({
        someObservable: di.inject(injectable),
        ...props,
      }),
    });

    const TestComponent = observer(() => (
      <SmartTestComponent data-some-prop-test={someObservable.get()} />
    ));

    const rendered = mount(<TestComponent />);

    act(() => {
      runInAction(() => {
        someObservable.set(42);
      });
    });

    expect(rendered.baseElement).toMatchSnapshot();
  });

  describe('given component, placeholder and async dependencies, when rendered', () => {
    let rendered;
    let asyncDependencyMock;

    beforeEach(async () => {
      asyncDependencyMock = asyncFn();

      const injectable = getInjectable({
        id: 'some-injectable-id',

        lifecycle: lifecycleEnum.transient,

        instantiate: () => asyncDependencyMock(),
      });

      di.register(injectable);

      const DumbTestComponent = ({ someDependency, ...props }) => (
        <div data-testid="some-dumb-test-component" {...props}>
          Some content: "{someDependency}"
        </div>
      );

      const SmartTestComponent = withInjectables(DumbTestComponent, {
        getProps: async (di, props) => ({
          someDependency: await di.inject(injectable),
          ...props,
        }),

        getPlaceholder: () => <div data-testid="some-placeholder" />,
      });

      rendered = mount(<SmartTestComponent data-some-prop-test />);
    });

    it('renders as placeholder', () => {
      expect(rendered.baseElement).toMatchSnapshot();
    });

    it('does not render component yet', () => {
      expect(
        rendered.queryByTestId('some-dumb-test-component'),
      ).not.toBeInTheDocument();
    });

    it('has placeholder', () => {
      expect(rendered.queryByTestId('some-placeholder')).toBeInTheDocument();
    });

    describe('when the dependency resolves', () => {
      beforeEach(async () => {
        await act(async () => {
          await asyncDependencyMock.resolve('some-async-value');
        });
      });

      it('renders', () => {
        expect(rendered.baseElement).toMatchSnapshot();
      });

      it('has component with async content', () => {
        expect(rendered.baseElement).toHaveTextContent(
          'Some content: "some-async-value"',
        );
      });

      it('has component', () => {
        expect(
          rendered.queryByTestId('some-dumb-test-component'),
        ).toBeInTheDocument();
      });

      it('no longer has placeholder', () => {
        expect(
          rendered.queryByTestId('some-placeholder'),
        ).not.toBeInTheDocument();
      });
    });
  });

  describe('given component with props, placeholder and async dependencies, when rendered', () => {
    let rendered;
    let asyncDependencyMock;

    beforeEach(async () => {
      asyncDependencyMock = asyncFn();

      const injectable = getInjectable({
        id: 'some-injectable-id',

        lifecycle: lifecycleEnum.transient,

        instantiate: () => asyncDependencyMock(),
      });

      di.register(injectable);

      const DumbTestComponent = ({ someDependency, ...props }) => (
        <div data-testid="some-dumb-test-component" {...props}>
          Some content: "{someDependency}"
        </div>
      );

      const SmartTestComponent = withInjectables(DumbTestComponent, {
        getProps: async (di, props) => ({
          someDependency: await di.inject(injectable),
          ...props,
        }),

        getPlaceholder: props => (
          <div data-testid={`some-placeholder-with-props(${props.someProp})`} />
        ),
      });

      rendered = mount(
        <SmartTestComponent data-some-prop-test someProp="some-prop-value" />,
      );
    });

    it('renders as placeholder using the props', () => {
      expect(rendered.baseElement).toMatchSnapshot();
    });

    it('has placeholder using props', () => {
      expect(
        rendered.queryByTestId('some-placeholder-with-props(some-prop-value)'),
      ).toBeInTheDocument();
    });
  });

  describe('given component, no placeholder and async dependencies, when rendered', () => {
    let rendered;
    let asyncDependencyMock;

    beforeEach(async () => {
      asyncDependencyMock = asyncFn();

      const injectable = getInjectable({
        id: 'some-injectable-id',

        lifecycle: lifecycleEnum.transient,

        instantiate: () => asyncDependencyMock(),
      });

      di.register(injectable);

      const DumbTestComponent = ({ someDependency, ...props }) => (
        <div {...props}>Some content: "{someDependency}"</div>
      );

      const SmartTestComponent = withInjectables(DumbTestComponent, {
        getProps: async (di, props) => ({
          someDependency: await di.inject(injectable),
          ...props,
        }),
      });

      rendered = mount(<SmartTestComponent data-some-prop-test />);
    });

    it('renders as null', () => {
      expect(rendered.container).toBeEmptyDOMElement();
    });

    describe('when the dependency resolves', () => {
      beforeEach(async () => {
        await act(() => asyncDependencyMock.resolve('some-async-value'));
      });

      it('renders', () => {
        expect(rendered.baseElement).toMatchSnapshot();
      });

      it('has component with async content', () => {
        expect(rendered.baseElement).toHaveTextContent(
          'Some content: "some-async-value"',
        );
      });
    });
  });

  it('given class component with sync dependencies, when rendered with ref, forwards ref', () => {
    const injectable = getInjectable({
      id: 'some-injectable-id',

      lifecycle: lifecycleEnum.transient,

      instantiate: () => 'some-injectable-value',
    });

    di.register(injectable);

    class DumbTestComponent extends React.Component {
      someProperty = 'some-property-accessed-with-ref';

      render() {
        const { someDependency, ...props } = this.props;

        return <div {...props}>Some content: "{someDependency}"</div>;
      }
    }

    const SmartTestComponent = withInjectables(DumbTestComponent, {
      getProps: (di, props) => ({
        someDependency: di.inject(injectable),
        ...props,
      }),
    });

    const ref = React.createRef();

    mount(<SmartTestComponent ref={ref} data-some-prop-test />);

    expect(ref.current.someProperty).toBe('some-property-accessed-with-ref');
  });

  it('given class component and async dependencies, when rendered with ref', async () => {
    const asyncDependencyMock = asyncFn();

    const injectable = getInjectable({
      id: 'some-injectable-id',

      lifecycle: lifecycleEnum.transient,

      instantiate: () => asyncDependencyMock(),
    });

    di.register(injectable);

    class DumbTestComponent extends React.Component {
      someProperty = 'some-property-accessed-with-ref';

      render() {
        const { someDependency, ...props } = this.props;

        return <div {...props}>Some content: "{someDependency}"</div>;
      }
    }

    const SmartTestComponent = withInjectables(DumbTestComponent, {
      getProps: async (di, props) => ({
        someDependency: await di.inject(injectable),
        ...props,
      }),
    });

    const ref = React.createRef();

    mount(<SmartTestComponent ref={ref} data-some-prop-test />);

    await act(() => asyncDependencyMock.resolve('some-async-value'));

    expect(ref.current.someProperty).toBe('some-property-accessed-with-ref');
  });

  it('given component, props and a dependency using instantiation parameter, when rendered, renders with the dependency having props as instantiation parameter', () => {
    const injectable = getInjectable({
      id: 'some-injectable-id',

      lifecycle: lifecycleEnum.transient,

      instantiate: (_, instantiationParameter) =>
        `some-injectable-value: ${instantiationParameter}`,
    });

    di.register(injectable);

    const DumbTestComponent = ({ someDependency, ...props }) => (
      <div {...props}>Some content: "{someDependency}"</div>
    );

    const SmartTestComponent = withInjectables(DumbTestComponent, {
      getProps: (di, { someInstantiationParameter, ...props }) => ({
        someDependency: di.inject(injectable, someInstantiationParameter),

        ...props,
      }),
    });

    const rendered = mount(
      <SmartTestComponent
        data-some-prop-test
        someInstantiationParameter="some-instantiation-parameter-value"
      />,
    );

    expect(rendered.baseElement).toMatchSnapshot();
  });
});

class ErrorBoundary extends React.Component {
  constructor(props) {
    super(props);
    this.state = { hasError: false };
  }

  static getDerivedStateFromError() {
    return { hasError: true };
  }

  componentDidCatch(error, errorInfo) {
    this.props.onError(error.message);
  }

  render() {
    if (this.state.hasError) {
      return null;
    }

    return this.props.children;
  }
}
