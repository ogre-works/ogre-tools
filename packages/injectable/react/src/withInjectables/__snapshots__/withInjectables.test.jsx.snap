// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withInjectables given anonymous component, injecting a token, when rendered, works 1`] = `
<body>
  <div>
    irrelevant
  </div>
</body>
`;

exports[`withInjectables given component and sync dependencies, when rendered without provider, renders to null 1`] = `
<body>
  <div />
</body>
`;

exports[`withInjectables given component and sync dependencies, when rendered, renders with dependencies 1`] = `
<body>
  <div>
    <div
      data-some-prop-test="true"
    >
      Some content: "
      some-injectable-value
      "
    </div>
  </div>
</body>
`;

exports[`withInjectables given component with props, placeholder and async dependencies, when rendered renders as placeholder using the props 1`] = `
<body>
  <div>
    <div
      data-testid="some-placeholder-with-props(some-prop-value)"
    />
  </div>
</body>
`;

exports[`withInjectables given component, and rendered, when re-rendered, works 1`] = `
<body>
  <div>
    irrelevant
  </div>
</body>
`;

exports[`withInjectables given component, no placeholder and async dependencies, when rendered when the dependency resolves renders 1`] = `
<body>
  <div>
    <div
      data-some-prop-test="true"
    >
      Some content: "
      some-async-value
      "
    </div>
  </div>
</body>
`;

exports[`withInjectables given component, placeholder and async dependencies, when rendered renders as placeholder 1`] = `
<body>
  <div>
    <div
      data-testid="some-placeholder"
    />
  </div>
</body>
`;

exports[`withInjectables given component, placeholder and async dependencies, when rendered when the dependency resolves renders 1`] = `
<body>
  <div>
    <div
      data-some-prop-test="true"
      data-testid="some-dumb-test-component"
    >
      Some content: "
      some-async-value
      "
    </div>
  </div>
</body>
`;

exports[`withInjectables given component, props and a dependency using instantiation parameter, when rendered, renders with the dependency having props as instantiation parameter 1`] = `
<body>
  <div>
    <div
      data-some-prop-test="true"
    >
      Some content: "
      some-injectable-value: some-instantiation-parameter-value
      "
    </div>
  </div>
</body>
`;
